packages/eslint-plugin/src/rules/dependency-checks.ts

```diff
@@ -0,0 +1,363 @@
+  | 'obsoleteDependency'
+  | 'versionMismatch'
+  | 'missingDependencySection';
+
+export const RULE_NAME = 'dependency-checks';
+
+export default createESLintRule<Options, MessageIds>({
+  name: RULE_NAME,
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: `Checks dependencies in project's package.json for version mismatches`,
+      recommended: 'error',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          buildTargets: [{ type: 'string' }],
+          ignoreDependencies: [{ type: 'string' }],
+          checkMissingDependencies: { type: 'boolean' },
+          checkObsoleteDependencies: { type: 'boolean' },
+          checkVersionMismatches: { type: 'boolean' },
+          checkMissingPackageJson: { type: 'boolean' },
+        },
+        additionalProperties: false,
+      },
+    ],
+    messages: {
+      noPackageJson: `A "package.json" file is required for buildable libraries, but does not exist in "{{projectName}}".`,
+      missingDependency: `The "{{projectName}}" uses the package "{{packageName}}", but it is missing from the project's "package.json".`,
+      obsoleteDependency: `The "{{packageName}}" package is found in the "package.json" but is not used by "{{projectName}}".`,
+      versionMismatch: `The "{{packageName}}" package is listed in "{{projectName}}"'s "package.json" but it's range is not compatible with the installed version. Installed version: {{version}}.`,
+      missingDependencySection: `Dependencies section is missing from the "package.json".`,
+    },
+  },
+  defaultOptions: [
+    {
+      buildTargets: ['build'],
+      checkMissingDependencies: true,
+      checkObsoleteDependencies: true,
+      checkVersionMismatches: true,
+      checkMissingPackageJson: true,
+      ignoredDependencies: [],
+    },
+  ],
+  create(
+    context,
+    [
+      {
+        buildTargets,
+        ignoredDependencies,
+        checkMissingDependencies,
+        checkObsoleteDependencies,
+        checkVersionMismatches,
+        checkMissingPackageJson,
+      },
+    ]
+  ) {
+    if (!(context.parserServices as any).isJSON) {
+      return {};
+    }
+    const fileName = normalizePath(context.getFilename());
+    // support only package.json and project.json
+    if (
+      !fileName.endsWith('/package.json') &&
+      !fileName.endsWith('/project.json')
+    ) {
+      return {};
+    }
+
+    const projectPath = normalizePath(globalThis.projectPath || workspaceRoot);
+    const sourceFilePath = getSourceFilePath(fileName, projectPath);
+    const { projectGraph, projectRootMappings, projectFileMap } =
+      readProjectGraph(RULE_NAME);
+
+    if (!projectGraph) {
+      return {};
+    }
+
+    const sourceProject = findProject(
+      projectGraph,
+      projectRootMappings,
+      sourceFilePath
+    );
+
+    // check if source project exists, it's a library
+    if (!sourceProject || sourceProject.type !== 'lib') {
```

Comment:
```suggestion
    if (!sourceProject) {
```

I wouldn't worry about what type the project is. As long as it has a `package.json` and a build target, this rule should work.